File: workflow-booking/src/main/java/com/acme/workflows/booking/java/BookingWorkflowImpl.java
package com.acme.workflows.booking.java;

import com.acme.services.notification.BookingRequest;
import com.acme.services.notification.NotificationService;
import com.acme.utils.AbstractWorkflow;
import com.acme.workflows.booking.BookingRequestStatus;
import com.acme.workflows.booking.BookingWorkflow;
import com.acme.workflows.payment.PaymentWorkflow;
import io.infinitic.annotations.Ignore;
import io.infinitic.workflows.Channel;
import io.infinitic.workflows.Deferred;

import java.time.Duration;
import java.time.Instant;

import static io.infinitic.workflows.DeferredKt.or;

public class BookingWorkflowImpl extends AbstractWorkflow implements BookingWorkflow {
    @Ignore
    private final long DAY_IN_SECONDS = 10L; // 24 * 3600

    // create channel for BookingRequestStatus
    private final Channel<BookingRequestStatus> responseChannel = channel();

    // create stub for HostService
    private final NotificationService notificationService = newService(NotificationService.class);

    // create stub for PaymentWorkflow
    private final PaymentWorkflow paymentWorkflow = newWorkflow(PaymentWorkflow.class);

    @Override
    public Channel<BookingRequestStatus> getResponseChannel() {
        return responseChannel;
    }

    @Override
    public void start(BookingRequest request) {
        // start accepting one signal in the channel
        Deferred<BookingRequestStatus> deferredResponse = responseChannel.receive(1);

        for (int attempt = 1; attempt <= 3; attempt++) {
            // notify host of traveler request
            dispatchVoid(notificationService::notifyHostOfRequest, request, attempt);

            // start a timer for a day
            Deferred<Instant> timer = timer(Duration.ofSeconds(DAY_IN_SECONDS));

            // Checks the completion of one of them
            or(timer, deferredResponse).await();

            // exit loop if we received a response from the host
            if (deferredResponse.isCompleted()) break;
        }

        // we did not receive host's response
        if (!deferredResponse.isCompleted()) {
            // notify host of request expiration
            dispatchVoid(notificationService::notifyHostOfRequestExpiration, request);
            // notify traveler of request expiration
            dispatchVoid(notificationService::notifyTravelerOfRequestExpiration, request);
            // workflow stops here
            return;
        }

        BookingRequestStatus response = deferredResponse.await();
        switch (response) {
            case ACCEPTED:
                bookingAccepted(request);
                break;
            case DENIED:
                bookingDenied(request);
                break;
        }
    }

    private void bookingAccepted(BookingRequest request) {
        log("Booking accepted by Host " + request.getHost());

        // trigger deposit workflow and wait for it
        paymentWorkflow.getDeposit(request);

        // notify host of the successful payment
        dispatchVoid(notificationService::notifyHostOfPaymentSuccess, request);

        // notify traveler of the successful booking
        dispatchVoid(notificationService::notifyTravelerOfBookingSuccess, request);
    }

    private void bookingDenied(BookingRequest request) {
        log("Booking denied by Host " + request.getHost());

        // notify traveler of host denial
        dispatchVoid(notificationService::notifyTravelerOfBookingDenial, request);
    }
}

File: workflow-booking/src/main/kotlin/com/acme/workflows/booking/kotlin/BookingWorkflowImpl.kt
package com.acme.workflows.booking.kotlin

import com.acme.contracts.services.notification.BookingRequest
import com.acme.contracts.services.notification.NotificationService
import com.acme.common.AbstractWorkflow
import com.acme.contracts.workflows.booking.BookingRequestStatus
import com.acme.contracts.workflows.booking.BookingWorkflow
import com.acme.contracts.workflows.payment.PaymentWorkflow
import io.infinitic.annotations.Ignore
import io.infinitic.cache.config.CaffeineCacheConfig
import io.infinitic.storage.compression.CompressionConfig
import io.infinitic.storage.config.MySQLStorageConfig
import io.infinitic.storage.config.StorageConfig.*
import io.infinitic.transport.config.PulsarTransportConfig
import io.infinitic.workers.InfiniticWorker
import io.infinitic.workers.config.ServiceTagEngineConfig
import io.infinitic.workflows.or
import java.time.Duration


class BookingWorkflowImpl: AbstractWorkflow(), BookingWorkflow {

    @Ignore
    private val DAY_IN_SECONDS = 10L // 24 * 3600

    // create stub for HostService
    private val notificationService = newService(NotificationService::class.java)

    // create stub for PaymentWorkflow
    private val paymentWorkflow: PaymentWorkflow = newWorkflow(
        PaymentWorkflow::class.java)

    // create channel for BookingRequestStatus
    private val responseChannel = channel<BookingRequestStatus>()

    override fun getResponseChannel() = responseChannel

    override fun start(request: BookingRequest) {
        // start accepting one signal in the channel
        val deferredResponse = responseChannel.receive(1)

        repeat(3) { attempt ->
            // notify host of traveler request
            dispatch(notificationService::notifyHostOfRequest, request, attempt + 1)
            // start a timer for a day
            val timer = timer(Duration.ofSeconds(DAY_IN_SECONDS))
            // wait for the timer or the signal
            (timer or deferredResponse).await()
            //  exit loop if we received a response from the host
            if (deferredResponse.isCompleted()) return@repeat
        }

        // we did not receive host's response
        if (!deferredResponse.isCompleted()) {
            // notify host of request expiration
            dispatch(notificationService::notifyHostOfRequestExpiration, request)
            // notify traveler of request expiration
            dispatch(notificationService::notifyTravelerOfRequestExpiration, request)
            // workflow stops here
            return
        }

        when(deferredResponse.await()) {
            BookingRequestStatus.ACCEPTED -> bookingAccepted(request)
            BookingRequestStatus.DENIED -> bookingDenied(request)
        }
    }

    private fun bookingAccepted(request: BookingRequest) {
        log("Booking accepted by Host ${request.host}")

        // trigger deposit workflow and wait for it
        paymentWorkflow.getDeposit(request)

        // notify host of the successful payment
        dispatch(notificationService::notifyHostOfPaymentSuccess, request)

        // notify traveler of the successful booking
        dispatch(notificationService::notifyTravelerOfBookingSuccess, request)
    }

    private fun bookingDenied(request: BookingRequest) {
        log("Booking denied by Host ${request.host}")

        // notify traveler of host denial
        dispatch(notificationService::notifyTravelerOfBookingDenial, request)
    }
}

fun main() {
    val transport = PulsarTransportConfig.builder()
        .setBrokerServiceUrl("pulsar://localhost:6650")
        .setWebServiceUrl("http://localhost:8080")
        .setTenant("infinitic")
        .setNamespace("dev")

    val storage: StorageConfigBuilder = MySQLStorageConfig.builder()
        .setCompression(CompressionConfig.bzip2)
        .setCache(
            CaffeineCacheConfig.builder().setExpireAfterWrite(600)
        )
        .setHost("localhost")
        .setPort(3306)
        .setDatabase("infinitic")
        .setUsername("root")
        .setPassword("***")

    val worker = InfiniticWorker.builder()
        .setTransport(transport)
        .addServiceTagEngine(
            ServiceTagEngineConfig.builder()
                .setServiceName("CarRentalService")
                .setConcurrency(5)
                .setStorage(storage)
        )
        .addServiceTagEngine(
            ServiceTagEngineConfig.builder()
                .setServiceName("FlightBookingService")
                .setConcurrency(5)
                .setStorage(storage)
        )
        .addServiceTagEngine(
            ServiceTagEngineConfig.builder()
                .setServiceName("HotelBookingService")
                .setConcurrency(5)
                .setStorage(storage)
        )
        .build()
}